scenario_09_sandbox.lua

```diff
@@ -15064,43 +15143,48 @@ function riptideBinarySector()
 	end
 
 	--- nebulas which are sucked from red giant
-	nebulaSpawner = function(self, obj, delta)
+	nebulaSpawner = function(obj, objectsRef)
 		local numNebulas = irandom(1, 3)
 		for i=1, numNebulas do
 			local sx, sy = riptideGamma:getPosition()
 
-			local angleToCenter = angleFromVectorNorth(sx, sy, centerX, centerY)
-			local dpx, dpy = vectorFromAngle(angleToCenter,1000)
-
-			local neb = Nebula():setPosition(sx + dpx + irandom(-1000, 1000), sy + dpy + irandom(-1000, 1000))
+			local angleToCenter = angleFromVectorNorth(centerX, centerY, sx, sy)
+			local dpx, dpy = vectorFromAngleNorth(angleToCenter, 1000)
 
+			local neb = Nebula():setPosition(sx + dpx + irandom(-500, 500), sy + dpy + irandom(-500, 500))
+			neb.facingAngle = irandom(0, 360)
 
 			local thisGravAccel = gravityAcceleration * random(80, 120) -- little bit of jitter so it spreads out nicely
 			--- some tweaking here to get a nice spiral
-			blackHoleOrbitUpdater(neb, centerX, centerY, slowOrbitDegPerSec, fastOrbitDegPerSec * 0.8, 10000, riptideGammaOrbitRadius, thisGravAccel)	
+			blackHoleOrbitUpdater(neb, centerX, centerY, slowOrbitDegPerSec, fastOrbitDegPerSec, 10000, riptideGammaOrbitRadius, thisGravAccel)
 			update_system:addUpdate(neb, "nebula misc", {
 				nextExplosionAt=0,
 				update=nebulaRotationAndFrictionUpdater
 			})
 
-			table.insert(objects, neb)
+			table.insert(objectsRef, neb)
 			neb:onDestroyed(function(obj, instigator)
 				--- remove self from objects
-				for i=1, #objects do
-					if obj == objects[i] then
-						table.remove(objects, i)
+				-- well, not really remove, because it messes up the destroy hook of the region
+				-- but set as nil to conserve memory (I hope)
+				for i=1, #objectsRef do
+					if obj == objectsRef[i] then
+						objectsRef[i] = nil
 						break
 					end
 				end
 			end)
 		end
 	end
-	update_system:addPeriodicCallback(riptideGamma, nebulaSpawner, 5, 0, 2)
+	update_system:addPeriodicCallback(riptideGamma, nebulaSpawner, 5, 0, 2, objects)
```

Comment:
This must be where that generic 'data' was used that I asked about earlier. @StarryWisdom knows the update system better than I do, so I'd like to hear from him. I generally like things to be spelled out rather than named something generic like 'data.' If there's a good reason, there should at least be some comments describing what's going on

Reply:
addPeriodicCallback doesn't seem to allow for custom data to be passed, unlike addUpdate. I'm aware this solution isn't elegant and would like to hear your guys' thoughts

Reply:
the data argument is likely to be an issue here in the long term (I dont want to stop merging if it solves real problems - I will fix it later if needed)
I am slightly concerned I am missing lifetime issues but
update_system:addPeriodicCallback(riptideGamma, function(obj) nebulaSpawner(obj,objects) end, 5, 0, 2)
I think should work while not requiring more parameters on addPeriodicCallback

There may be a neater soultion embedding objects into riptideGamma (which in turn may need to be turned into a newPhonySpaceObject to prevent EE :destroying() it)

The data for addUpdate was intended to be an internal detail (mostly anyway) to update_system - the intent was a lua equivilant to C++ having a virtual base class with an update function, this has somewhat proved a mistake on my part, I'm not keen on spreading it too far

the most important part though is - if it works and fixes things go for it, I may well back it out (and fix this part of the sandbox) when I get back round to it though

Reply:
> I am slightly concerned I am missing lifetime issues but

Xansta revealed some lifetime issues... but now that I'm testing again, everything seems fine. Therefore I will revert this change.